core/executor/src/wasm_executor.rs

```diff
@@ -679,6 +705,40 @@ impl WasmExecutor {
 		method: &str,
 		data: &[u8],
 	) -> Result<Vec<u8>> {
+		self.call_in_wasm_module_with_custom_signature(
+			ext,
+			module_instance,
+			method,
+			|alloc| {
+				let offset = alloc(data)?;
+				Ok(vec![I32(offset as i32), I32(data.len() as i32)])
+			},
+			|res, memory| {
+				if let Some(I64(r)) = res {
+					let offset = r as u32;
+					let length = (r >> 32) as usize;
```

Comment:
Hm, I wonder if this is correct. Can you check my logic?

1. `r` is [`i64`](https://paritytech.github.io/wasmi/wasmi/enum.RuntimeValue.html#variant.I64).
2. According to the reference, right shift is logical for unsinged values but arithmetic for signed values.
3. So `0x8000_0000_0000_0000 >> 32` is an arithmetic shift and will give you `0xFFFF_FFFF_8000_0000`.

Reply:
@tomusdrw FYI. We both need to go back and learn shifting :D